File: September 2022/September_15.cpp
class Solution {
public:
    vector<int> findOriginalArray(vector<int>& changed) {
        map<int, int> mp;
        for(int i : changed)
            mp[i]++;
        int n = changed.size();
        if((n & 1) == 1){
            vector<int> v;
            return v;
        }
        vector<int> ans (n / 2);
        int j = 0;
        sort(changed.begin(), changed.end());
        for(int i = 0; i < n; i++){
            if(mp.find(changed[i]) != mp.end() && mp.find(2 * changed[i]) != mp.end()){
                if(changed[i] == 0){
                    if(mp[0] > 2){
                        mp[0] -= 2;
                        ans[j++] = 0;
                    } else if(mp[0] == 2){
                        mp.erase(0);
                        ans[j++] = 0;
                    }
                } else{
                    if(mp[changed[i]] == 1)
                        mp.erase(changed[i]);
                    else
                        mp[changed[i]]--;
                    if(mp[2 * changed[i]] == 1)
                        mp.erase(2 * changed[i]);
                    else
                        mp[2 * changed[i]]--;
                    ans[j++] = changed[i];
                }
            }
        }
        if(j == n / 2)
            return ans;
        vector<int> v;
        return v;
    }
};

File: September 2022/September_15.java
class Solution {
    public int[] findOriginalArray(int[] changed) {
        Map<Integer, Integer> map = new HashMap<>();
        for(int i : changed)
            map.put(i, map.getOrDefault(i, 0) + 1);
        int n = changed.length;
        if((n & 1) == 1)
            return new int[0];
        int[] ans = new int[n / 2];
        int j = 0;
        Arrays.sort(changed);
        for(int i = 0; i < n; i++){
            if(map.containsKey(changed[i]) && map.containsKey(2 * changed[i])){
                if(changed[i] == 0){
                    if(map.get(0) > 2){
                        map.put(0, map.get(0) - 2);
                        ans[j++] = 0;
                    } else if(map.get(0) == 2){
                        map.remove(0);
                        ans[j++] = 0;
                    }
                } else{
                    if(map.get(changed[i]) == 1)
                        map.remove(changed[i]);
                    else
                        map.put(changed[i], map.get(changed[i]) - 1);
                    if(map.get(2 * changed[i]) == 1)
                        map.remove(2 * changed[i]);
                    else
                        map.put(2 * changed[i], map.get(2 * changed[i]) - 1);
                    ans[j++] = changed[i];
                }
            }
        }
        if(j == n / 2)
            return ans;
        return new int[0];
    }
}
